pkg/sql/opt/memo/memo.go

```diff
@@ -341,58 +341,85 @@ func (m *Memo) LookupPrivate(id PrivateID) interface{} {
 // --------------------------------------------------------------------
 
 func (m *Memo) String() string {
+	return m.Display(GroupID(0))
+}
+
+// Display writes a memo to a human-readable format.
+func (m *Memo) Display(rootGroup GroupID) string {
```

Comment:
I think this logic has gotten to a point of complexity and volume where you should break it out into its own file and helper struct:
```
type memoFormat struct {
  mem *Memo
  buf bytes.Buffer
  numbering []GroupID
  ... // whatever else you need
}
```
Any time you start to see code with lots of helper methods and many parameters, it's a good sign that a helper struct is in order (usually allocated on the stack).